**Απόστολος Μουστάκης**

**AM: 03400182**

<font color='darkred'>Οι απαντήσεις σε όλα τα ερωτήματα γράφονται με αυτό το χρώμα ώστε να ξεχωρίζουν από τις εκφωνήσεις </font>

<h1><b>Markov Decision Processes</h1></b>
<p align="justify">Στη συγκεκριμένη άσκηση θα μελετήσετε τους αλγορίθμους <i>Policy Iteration</i> και <i>Value Iteration</i>, καθώς και θα εξοικειωθείτε με βασικές έννοιες των <i>Markov Decision Processes</i>. Οι αλγόριθμοι <i>Policy Iteration</i> και <i>Value Iteration</i> είναι από τους βασικούς αλγορίθμους δυναμικού προγραμματισμού που χρησιμοποιούνται για την επίλυση της εξίσωσης <i>Bellman</i> σε <i>Markov Decision Processes</i>.</p> 
<p align="justify">Το πρόβλημα που θα μελετήσετε είναι αυτό της παγωμένης λίμνης (Frozen Lake) με μέγεθος πλέγματος 8 x 8.</p>


<h2><b>Εξοικείωση με τη βιβλιοθήκη <i>Gym</i></b></h2>

In [ ]:
!pip uninstall -y gym
!pip install setuptools==66
!pip install gym==0.18.0

In [ ]:
import numpy as np
import gym
from gym import wrappers

Με την παρακάτω εντολή, ορίζετε το πρόβλημα που θα μελετηθεί:

In [ ]:
env_name = 'FrozenLake8x8-v0'
env = gym.make(env_name)

Με τις παρακάτω εντολές, θα επαναφέρετε τον Agent στην αρχική του θέση και θα οπτικοποιήσετε το πλέγμα και τη θέση του Agent

In [ ]:
env.reset()
env.render()


SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG


Με τις παρακάτω εντολές, ορίζετε την επόμενη ενέργεια με τυχαίο τρόπο και ο Agent κάνει ένα βήμα.

In [ ]:
next_action = env.action_space.sample()
env.step(next_action)
env.render()

  (Left)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG


Να εκτελέσετε αρκετές φορές τις τελευταίες εντολές και να παρατηρήσετε κάθε φορά την ενέργεια που ζητείται από τον agent να εκτελέσει και την ενέργεια που αυτός πραγματοποιεί. Πραγματοποιεί πάντοτε ο agent την κίνηση που του ζητείται; Είναι ντετερμινιστικές ή στοχαστικές οι κινήσεις του agent;

<font color='darkred'>Εκτελώντας τις τελευταίες εντολές αρκετές φορές παρατηρώ πως ο πράκτορας (agent) δεν εκτελεί πάντα την κίνηση που του ζητείται. Συνεπώς οι κινήσεις του είναι στοχαστικές και όχι ντετερμινιστικές. Ο λόγος που συμβαίνει αυτό είναι διότι υπάρχουν δύο εκδόσεις του προβλήματος της Παγωμένης Λίμνης (Frozen Lake). Η μία έκδοση έχει ολισθηρό πάγο (slippery ice), όπου οι επιλεγμένες ενέργειες έχουν τυχαία πιθανότητα να αγνοηθούν από τον πράκτορα, ενώ η άλλη έχει μη ολισθηρό πάγο (non-slippery ice), όπου οι επιλεγμένες ενέργειες πραγματοποιούνται πάντα από τον πράκτορα. Η παραδοχή αυτή έγινε στον κώδικα όταν φτιάξαμε το περιβάλλον με την βιβλιοθήκη gym. Αν θέλαμε την δεύτερη περίπτωση τότε αρκεί η εξής μικρή μεταποίηση στον κώδικα: `env = gym.make(env_name, is_slippery=False)`.
Περισσότερα για το πρόβλημα της Παγωμένης Λίμνης θα αναλυθούν στο επόμενο ερώτημα. 


<h2><b>Ερωτήσεις</b></h2>

<li>Μελετώντας <a href="https://machinelearningjourney.com/index.php/2020/07/02/frozenlake/">αυτό</a> και βασισμένοι στα συμπεράσματα του προηγούμενου ερωτήματος να περιγράψετε σύντομα το πρόβλημα της παγωμένης λίμνης ως πρόβλημα βελτιστοποίησης. Ποιος είναι ο στόχος του agent;</li>

<font color='darkred'>Το πρόβλημα της Παγωμένης Λίμνης (Frozen Lake) αφορά ένα περιβάλλον, το οποίο αποτελεί ένα πλέγμα με “πλακίδια” (tiles). Στην συγκεκριμένη άσκηση το πλέγμα είναι 8x8, δηλαδή διαθέτει 64 tiles, αλλά προφανώς μπορεί να είναι και διαφορετικών διαστάσεων. Στο περιβάλλον αυτό υπάρχει ο πράκτορας (agent), ο οποίος θέλει να μετακινηθεί από το αρχικό πλακίδιο (S: Starting point) στο πλακίδιο στόχο (G: Goal), διασχίζοντας το πλέγμα, δηλαδή την λίμνη, με ασφάλεια. Τα ενδιάμεσα πλακίδια είναι είτε παγωμένη επιφάνεια (F: Frozen surface) είτε τρύπες (H: Hole). Για να διασχίσει λοιπόν με ασφάλεια την λίμνη ο πράκτορας πρέπει να περνάει μόνο από παγωμένη επιφάνεια, η οποία είναι ασφαλής. Αν σε οποιοδήποτε σημείο της διαδρομής πέσει σε τρύπα τότε μένει εκεί για πάντα. Οι κινήσεις που μπορεί να κάνει είναι πάνω, κάτω, δεξιά και αριστερά. Όπως περιέγραψα προηγουμένως στην εκδοχή αυτή το πλακίδιο είναι ολισθηρό, το οποίο σημαίνει ότι ενώ ο πράκτορας μπορεί να θέλει να πάει κάτω, υπάρχει πιθανότητα να «γλιστρήσει» και να πάει για παράδειγμα αριστερά όπου εκεί μπορεί να βρίσκεται μία τρύπα. Στόχος λοιπόν του πράκτορα είναι να μεταβεί από το αρχικό πλακίδιο στο τελικό πλακίδιο χωρίς να πέσει σε κάποια τρύπα. Το πρόβλημα αυτό αποτελεί πρόβλημα βελτιστοποίησης, καθώς η μετάβαση αυτή πρέπει να γίνει με τον μικρότερο αριθμό βημάτων/ δράσεων του πράκτορα. Αξίζει τέλος να σημειωθεί πως μόνο το πλακίδιο στόχος έχει επιβράβευση +1, ενώ όλα τα υπόλοιπα πλακίδια έχουν επιβράβευση 0. 

<li>Να διατυπώσετε την ιδιότητα <i>Markov</i>. Πώς απλοποιεί η ιδιότητα αυτή τη μελέτη του συγκεκριμένου προβλήματος;</li>

<font color='darkred'>Έστω $(X_n) = (X_0, X_1, X_2, ...)$ μία στοχαστική διαδικασία σε διακριτό χρόνο $n = 0,1,2,...$ και διακριτό χώρο $S$. Λέμε οτι για την $(X_n)$ ισχύει η ιδιότητα Markov όταν για όλους τους χρόνους $n$ και για όλες τις καταστάσεις $(x_0, x_1,...,x_n, x_{n+1})∈S$ ισχύει:
$$ P(X_{n+1}=x_{n+1} \mid X_{n}=x_{n}, X_{n-1} = x_{n-1}, \dots,X_0=x_0) = P(X_{n+1}=x_{n+1} \mid X_{n}=x_{n}) $$

<font color='darkred'>Η ιδιότητα Markov εκφράζει το γεγονός ότι σε ένα δεδομένο χρονικό βήμα και γνωρίζοντας την τρέχουσα κατάσταση, δεν θα λάβουμε πρόσθετες πληροφορίες για το μέλλον συλλέγοντας πληροφορίες για το παρελθόν. Με άλλα λόγια μας ενδιαφέρει μόνο το παρόν και όχι το παρελθόν.

<font color='darkred'>Η ιδιότητα αυτή απλοποιεί το συγκεκριμένο πρόβλημα, καθώς πλέον το πρόβλημα μοντελοποιείται ως MDP (Markov Decision Process) και συνεπώς μπορούν να χρησιμοποιηθούν τεχνικές και αλγόριθμοι δυναμικού προγραμματισμού για την επίλυση του. Η απλοποίηση έγκειται στο γεγονός πως το ευρύτερο πρόβλημα ανάγεται σε πολλά μικρά προβλήματα όπου κάθε φορά μας ενδιαφέρει το επόμενο βήμα του πράκτορα σε σχέση με την θέση που βρίσκεται (στην οποία κατέληξε κάνοντας το προηγούμενο βήμα), χωρίς να μας ενδιαφέρουν τα προηγούμενα βήματα που έκανε, δηλαδή ο τρόπος με τον οποίο κατέληξε στην θέση αυτή.

<li>Να περιγράψετε σύντομα τους αλγορίθμους <i>Policy Iteration</i> και <i>Value Iteration</i>, δίνοντας έμφαση στο διαφορετικό τρόπο με τον οποίο προσεγγίζουν την επίλυση του προβλήματος. Είναι εγγυημένο ότι οι δύο αλγόριθμοι θα συγκλίνουν στη βέλτιστη πολιτική; Αν ναι, οδηγούν σε ίδια ή διαφορετική βέλτιστη πολιτική;</li>

<font color='darkred'>Οι Value Iteration και Policy Iteration είναι αλγόριθμοι οι οποίοι χρησιμοποιούνται για την επίλυση προβλημάτων που είναι MDPs (Markov Decision Processes) υπολογίζοντας την καλύτερη πολιτική $π’(s)$.

<font color='darkred'>Ο **Value iteration** χωρίζεται σε δύο βήματα: εύρεση των καλύτερων values (επαναληπτική διαδικασία) και εξαγωγή πολιτικής (μία φορά). Πιο συγκεκριμένα ξεκινάμε με μία αυθαίρετη value function και επαναληπτικά βρίσκουμε τα καλύτερα values. Σε κάθε βήμα τα values μετακινούνται πιο κοντά στα βέλτιστα values. Αν $S$ είναι ο χώρος των καταστάσεων και $A$ ο χώρος των δράσεων τότε κάθε επανάληψη παίρνει $Ο(S^2A)$. Αφού βρούμε τα βέλτιστα values κάνουμε εξαγωγή πολιτικής μία μόνο φορά και με αυτόν τον τρόπο βρίσκουμε την καλύτερη πολιτική. Μαθηματικά ο αλγόριθμος είναι ο εξής:

1.  Αρχικοποίησε $V_0(s) = 0$ για όλες τις καταστάσεις $s$
2.  Υπολόγισε νέα values για όλες τις καταστάσεις s ως: $$V_{k+1}(s) = \max_a \sum_{s'} T(s,a,s')[R(s,a,s')+γV_k(s')]$$
1.  Επανέλαβε το βήμα 2 μέχρι να επέλθει σύγκλιση

<font color='darkred'>Πρακτικά χρησιμοποιούμε έναν πολύ μικρό αριθμό, έστω $ε$, για να καθορίσουμε πότε έχει συγκλίνει ο αλγόριθμος. Αν η διαφορά των values μεταξύ δύο επαναλήψεων είναι μικρότερη του $ε$ τότε ο αλγόριθμος έχει συγκλίνει. Αφού έχει συγκλίνει εξάγουμε την βέλτιστη πολιτική ως:  $$π'(s) = arg\max_a \sum_{s'} T(s,a,s')[R(s,a,s')+γV'(s')]$$

<font color='darkred'>Ο **Policy Iteration** χωρίζεται σε δύο βήματα: Policy Evaluation (επαναληπτική διαδικασία) και Policy Improvement (επαναληπτική διαδικασία). Πιο συγκεκριμένα με τον Policy Iteration ξεκινάμε με μια αυθαίρετη πολιτική και την βελτιώνουμε επαναληπτικά μέχρι να βρούμε την βέλτιστή. Στο βήμα Policy Evaluation διορθώνουμε την πολιτική και υπολογίζουμε τις τιμές κάθε κατάστασης (state) στο πλαίσιο της πολιτικής αυτή. Έπειτα για Policy Improvement κάνουμε update την πολιτική χρησιμοποιώντας έναν αλγόριθμο παρόμοιο με τον αλγόριθμο εξαγωγής πολιτικής που αναφέρθηκε προηγουμένως, με την διαφορά ότι ελέγχουμε αν η πολιτική έχει συγκλίνει ή όχι, δηλαδή αν έχει αλλάξει η όχι. Αν η πολιτική έχει αλλάξει τότε επιστρέφουμε στο βήμα Policy Evaluation και συνεχίζουμε την διαδικασία αυτή μέχρι να συγκλίνει η πολιτική, δηλαδή να μην αλλάξει από το πιο πρόσφατο βήμα. Αν $S$ είναι ο χώρος των καταστάσεων και $A$ ο χώρος των δράσεων σε αυτή την περίπτωση κάθε επανάληψη παίρνει $Ο(A^2)$. Μαθηματικά ο αλγόριθμος είναι ο εξής:

<font color='darkred'>Για Policy Evaluation:
1.  Αρχικοποίησε $V_0(s) = 0$ για όλες τις καταστάσεις $s$
2. Υπολόγισε νέα values για όλες τις καταστάσεις s δεδομένης πολιτικής π ως: $$V_{k+1}(s) = \sum_{s'} T(s,π(s),s')[R(s,π(s),s')+γV_k(s')]$$
1.  Επανέλαβε το βήμα 2 μέχρι να επέλθει σύγκλιση

<font color='darkred'>Για Policy Improvement:
1.  Ενημέρωσε την πολιτική ως: $$π(s) = arg\max_a \sum_{s'} T(s,a,s')[R(s,a,s')+γV'(s')]$$
2. Επανέλαβε το βήμα 1 μέχρι να επέλθει σύγκλιση

<font color='darkred'>Όπως ήδη ανέφερα οι δύο αλγόριθμοι χρησιμοποιούνται για προβλήματα, τα οποία είναι Markov Decision Processes. Συνεπώς είναι εγγυημένο πως και οι δύο θα καταλήξουν σε βέλτιστη πολιτική. Αν η βέλτιστη πολιτική είναι μόνο μία τότε θα καταλήξουν στην ίδια. Αν υπάρχουν περισσότερες από μία βέλτιστες πολιτικές τότε μπορεί να καταλήξουν σε διαφορετική.


<li>Να εκτελέσετε τα προγράμματα που σας δίνονται, τα οποία επιλύουν το
πρόβλημα της παγωμένης λίμνης, χρησιμοποιώντας τους αλγορίθμους <i>Policy
Iteration</i> και <i>Value Iteration</i> αντίστοιχα. Ποια μέθοδος συγκλίνει στη βέλτιστη λύση σε λιγότερα βήματα; Τι συμπέρασμα βγάζετε; Να ελέγξετε το χρόνο εκτέλεσης του κάθε προγράμματος, χρησιμοποιώντας την εντολή <i>time</i>. Τι συμπέρασμα βγάζετε ως προς την πολυπλοκότητα του κάθε αλγορίθμου;</li>

<font color='darkred'>Παρακάτω εκτελώ τον κώδικα της άσκησης για την επίλυση του προβλήματος της Παγωμένης Λίμνης χρησιμοποιώντας τους αλγορίθμους Policy Iteration και Value Iteration. Ο χρόνος εκτέλεσης κάθε προγράμματος υπολογίζεται με την χρήση της `%%time`. 

<h2><b>Policy Iteration</b></h2>

In [ ]:
%%time
"""
Solving FrozenLake8x8 environment using Policy iteration.
Author : Moustafa Alzantot (malzantot@ucla.edu)
"""

import numpy as np
import gym
from gym import wrappers


def run_episode(env, policy, gamma = 1.0, render = False):
    """ Runs an episode and return the total reward """
    obs = env.reset()
    total_reward = 0
    step_idx = 0
    while True:
        if render:
            env.render()
        obs, reward, done , _ = env.step(int(policy[obs]))
        total_reward += (gamma ** step_idx * reward)
        step_idx += 1
        if done:
            break
    return total_reward


def evaluate_policy(env, policy, gamma = 1.0, n = 100):
    scores = [run_episode(env, policy, gamma, False) for _ in range(n)]
    return np.mean(scores)

def extract_policy(v, gamma = 1.0):
    """ Extract the policy given a value-function """
    policy = np.zeros(env.nS)
    for s in range(env.nS):
        q_sa = np.zeros(env.nA)
        for a in range(env.nA):
            q_sa[a] = sum([p * (r + gamma * v[s_]) for p, s_, r, _ in  env.P[s][a]])
        policy[s] = np.argmax(q_sa)
    return policy

def compute_policy_v(env, policy, gamma=1.0):
    """ Iteratively evaluate the value-function under policy.
    Alternatively, we could formulate a set of linear equations in iterms of v[s] 
    and solve them to find the value function.
    """
    v = np.zeros(env.nS)
    eps = 1e-10
    while True:
        prev_v = np.copy(v)
        for s in range(env.nS):
            policy_a = policy[s]
            v[s] = sum([p * (r + gamma * prev_v[s_]) for p, s_, r, _ in env.P[s][policy_a]])
        if (np.sum((np.fabs(prev_v - v))) <= eps):
            # value converged
            break
    return v

def policy_iteration(env, gamma = 1.0):
    """ Policy-Iteration algorithm """
    policy = np.random.choice(env.nA, size=(env.nS))  # initialize a random policy
    max_iterations = 200000
    gamma = 1.0
    for i in range(max_iterations):
        old_policy_v = compute_policy_v(env, policy, gamma)
        new_policy = extract_policy(old_policy_v, gamma)
        if (np.all(policy == new_policy)):
            print ('Policy-Iteration converged at step %d.' %(i+1))
            break
        policy = new_policy
    return policy


if __name__ == '__main__':
    env_name  = 'FrozenLake8x8-v0'
    env = gym.make(env_name)
    env = env.unwrapped
    optimal_policy = policy_iteration(env, gamma = 1.0)
    scores = evaluate_policy(env, optimal_policy, gamma = 1.0)
    print('Average scores = ', np.mean(scores))

Policy-Iteration converged at step 12.
Average scores =  1.0
CPU times: user 2.91 s, sys: 185 ms, total: 3.09 s
Wall time: 2.89 s


<h2><b>Value Iteration</b></h2>

In [ ]:
%%time
"""
Solving FrozenLake8x8 environment using Value-Itertion.
Author : Moustafa Alzantot (malzantot@ucla.edu)
"""
import numpy as np
import gym
from gym import wrappers


def run_episode(env, policy, gamma = 1.0, render = False):
    """ Evaluates policy by using it to run an episode and finding its
    total reward.
    args:
    env: gym environment.
    policy: the policy to be used.
    gamma: discount factor.
    render: boolean to turn rendering on/off.
    returns:
    total reward: real value of the total reward recieved by agent under policy.
    """
    obs = env.reset()
    total_reward = 0
    step_idx = 0
    while True:
        if render:
            env.render()
        obs, reward, done , _ = env.step(int(policy[obs]))
        total_reward += (gamma ** step_idx * reward)
        step_idx += 1
        if done:
            break
    return total_reward


def evaluate_policy(env, policy, gamma = 1.0,  n = 100):
    """ Evaluates a policy by running it n times.
    returns:
    average total reward
    """
    scores = [
            run_episode(env, policy, gamma = gamma, render = False)
            for _ in range(n)]
    return np.mean(scores)

def extract_policy(v, gamma = 1.0):
    """ Extract the policy given a value-function """
    policy = np.zeros(env.nS)
    for s in range(env.nS):
        q_sa = np.zeros(env.action_space.n)
        for a in range(env.action_space.n):
            for next_sr in env.P[s][a]:
                # next_sr is a tuple of (probability, next state, reward, done)
                p, s_, r, _ = next_sr
                q_sa[a] += (p * (r + gamma * v[s_]))
        policy[s] = np.argmax(q_sa)
    return policy


def value_iteration(env, gamma = 1.0):
    """ Value-iteration algorithm """
    v = np.zeros(env.nS)  # initialize value-function
    max_iterations = 100000
    eps = 1e-20
    for i in range(max_iterations):
        prev_v = np.copy(v)
        for s in range(env.nS):
            q_sa = [sum([p*(r + prev_v[s_]) for p, s_, r, _ in env.P[s][a]]) for a in range(env.nA)] 
            v[s] = max(q_sa)
        if (np.sum(np.fabs(prev_v - v)) <= eps):
            print ('Value-iteration converged at iteration# %d.' %(i+1))
            break
    return v


if __name__ == '__main__':
    env_name  = 'FrozenLake8x8-v0'
    gamma = 1.0
    env = gym.make(env_name)
    env = env.unwrapped
    optimal_v = value_iteration(env, gamma);
    policy = extract_policy(optimal_v, gamma)
    policy_score = evaluate_policy(env, policy, gamma, n=1000)
    print('Policy average score = ', policy_score)

Value-iteration converged at iteration# 2357.
Policy average score =  1.0
CPU times: user 2.54 s, sys: 212 ms, total: 2.76 s
Wall time: 2.57 s


<font color='darkred'>Με βάση τα αποτελέσματα παρατηρούμε πως ο αλγόριθμος Policy Iteration συγκλίνει στην βέλτιστη λύση σε μόλις 12 βήματα, ενώ ο αλγόριθμος Value Iteration σε 2357. Παρόλο που ο αλγόριθμος Value Iteration απαιτεί πολύ περισσότερες επαναλήψεις εωσότου να καταλήξει στην βέλτιστη λύση ο χρόνος εκτέλεσης είναι παρόμοιος και για τους δύο και κινείται στην ίδια τάξη μεγέθους. Συγκεκριμένα ο χρόνος εκτέλεσης του Policy Iteration είναι ελάχιστα μεγαλύτερος (Wall time Policy Iteration: 2.89s, Wall time Value Iteration: 2.57s). Τα αποτελέσματα αυτά συνάδουν με τις διαφορές των δύο αλγορίθμων που αναλύθηκαν παραπάνω, καθώς και τις πολυπλοκότητες τους. Ο Policy Iteration έχει σημαντικά πιο υψηλή πολυπλοκότητα από τον Value Iteration και για αυτό παρόλο που συγκλίνει σε ένα πολύ μικρό αριθμό βημάτων ο χρόνος εκτέλεσης παραμένει παρόμοιος. 